plugin/pkg/admission/runtimeclass/admission.go

```diff
@@ -0,0 +1,207 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package runtimeclass contains an admission controller for modifying and validating new Pods to
+// take RuntimeClass into account. For RuntimeClass definitions which describe an overhead associated
+// with running a pod, this admission controller will set the pod.Spec.Overhead field accordingly. This
+// field should only be set through this controller, so vaidation will be carried out to ensure the pod's
+// value matches what is defined in the coresponding RuntimeClass.
+package runtimeclass
+
+import (
+	"fmt"
+	"io"
+	v1beta1 "k8s.io/api/node/v1beta1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninitailizer "k8s.io/apiserver/pkg/admission/initializer"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	nodev1beta1listers "k8s.io/client-go/listers/node/v1beta1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	node "k8s.io/kubernetes/pkg/apis/node"
+	nodev1beta1 "k8s.io/kubernetes/pkg/apis/node/v1beta1"
+	"k8s.io/kubernetes/pkg/features"
+	//nodev1beta1 "k8s.io/api/node/v1beta1"
+)
+
+// PluginName indicates name of admission plugin.
+const PluginName = "RuntimeClass"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return NewRuntimeClass(), nil
+	})
+}
+
+// RuntimeClass is an implementation of admission.Interface.
+// It looks at all new pods and sets pod.Spec.Overhead if a RuntimeClass is specified which
+// defines an Overhead. TODO: add blurb about validation/rejection handling
+type RuntimeClass struct {
+	*admission.Handler
+	client kubernetes.Interface
+	lister nodev1beta1listers.RuntimeClassLister
+}
+
+var _ admission.MutationInterface = &RuntimeClass{}
+var _ admission.ValidationInterface = &RuntimeClass{}
+
+var _ genericadmissioninitailizer.WantsExternalKubeInformerFactory = &RuntimeClass{}
+
+func (r *RuntimeClass) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	runtimeClassInformer := f.Node().V1beta1().RuntimeClasses()
+	r.SetReadyFunc(runtimeClassInformer.Informer().HasSynced)
+	r.lister = runtimeClassInformer.Lister()
+}
+
+func (r *RuntimeClass) ValidateInitialization() error {
+	if r.lister == nil {
+		return fmt.Errorf("missing RuntimeClass lister")
+	}
+	if r.client == nil {
+		return fmt.Errorf("missing client")
+	}
+	return nil
+}
+
+// Admit makes an admission decision based on the request attributes
+func (r *RuntimeClass) Admit(attributes admission.Attributes, o admission.ObjectInterfaces) error {
+
+	// Ignore all calls to subresources or resources other than pods.
+	if len(attributes.GetSubresource()) != 0 || attributes.GetResource().GroupResource() != api.Resource("pods") {
```

Comment:
Sorry, I know I originally said to inline this, but now that it's used in 2 places I do think it should be a function to dedupe the logic.